src/main/java/io/druid/segment/loading/DataSegmentPuller.java

```diff
@@ -10,10 +13,49 @@
 {
   /**
    * Pull down segment files for the given DataSegment and put them in the given directory.
+   * Deprecated in favor of URI pulling when possible
    *
    * @param segment The segment to pull down files for
-   * @param dir The directory to store the files in
+   * @param dir     The directory to store the files in
+   *
    * @throws SegmentLoadingException if there are any errors
    */
+  @Deprecated
   public void getSegmentFiles(DataSegment segment, File dir) throws SegmentLoadingException;
+
+  /**
+   * Pull down segment files for the given DataSegment and put them in the given directory.
+   *
+   * @param segmentUri The URI for the segment
+   * @param dir        The directory to store the files in
+   *
+   * @throws SegmentLoadingException if there are any errors
+   */
+  public void getSegmentFiles(URI segmentUri, File dir) throws SegmentLoadingException;
+
+  /**
+   * Open up an input stream to the requested URI
+   *
+   * @param segmentUri The URI needed to open
+   *
+   * @return InputStream to the requested URI
+   *
+   * @throws SegmentLoadingException
+   */
+  public InputStream getUriInputStream(URI segmentUri) throws SegmentLoadingException;
+
+  /**
+   * Builds a Predicate<Throwable> to determine if an error should retry.
+   *
+   * @return Predicate<Throwable> which returns true if the implementation should retry
+   */
+  public Predicate<Throwable> buildShouldRetryPredicate();
```

Comment:
This is following the method which was previously used regarding retries. Only S3 segments had the logic though. I think the original intention was that some errors are recoverable and some are not, this method should be able to determine if an error from any particular retry loop is recoverable.

Samza does a similar approach in its [kafka producer](https://github.com/apache/samza/blob/master/samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala#L110)